overlord/standby/standby_test.go

```diff
@@ -111,3 +111,71 @@ func (s *standbySuite) TestCanStandbyWithOpinion(c *C) {
 	s.canStandby = false
 	c.Check(m.CanStandby(), Equals, false)
 }
+
+type opine func() bool
+
+func (f opine) CanStandby() bool {
+	return f()
+}
+
+func (s *standbySuite) TestStartChecks(c *C) {
+	n := 0
+	ch1 := make(chan struct{})
+	ch2 := make(chan struct{})
+
+	defer standby.MockStandbyWait(time.Millisecond)()
+	defer standby.MockStateRequestRestart(func(_ *state.State, t state.RestartType) {
+		c.Check(t, Equals, state.RestartSocket)
+		n++
+		<-ch2
+	})()
+
+	opinion := false
+	m := standby.New(s.state)
+	m.AddOpinion(opine(func() bool {
+		<-ch1
+		return opinion
+	}))
+
+	m.Start()
+	ch1 <- struct{}{}
+	c.Check(n, Equals, 0)
+	ch1 <- struct{}{}
+	c.Check(n, Equals, 0)
+
+	opinion = true
+	ch1 <- struct{}{}
+	ch2 <- struct{}{}
+	c.Check(n, Equals, 1)
+
+	m.Stop()
+}
+
+func (s *standbySuite) TestStopWaits(c *C) {
+	defer standby.MockStandbyWait(time.Millisecond)()
+	defer standby.MockStateRequestRestart(func(*state.State, state.RestartType) {})()
+
+	ch := make(chan struct{})
+	m := standby.New(s.state)
+	m.AddOpinion(opine(func() bool {
+		time.Sleep(200 * time.Millisecond)
+		return false
+	}))
+
+	m.Start()
+
+	// wait enough so the actual Opinionator is called
+	time.Sleep(10 * time.Millisecond)
```

Comment:
I think we should sync with the opinionator being called and make sure that Stop actually completes. Otherwise this feels like the issue I tried to fix in #6041. I tried the diff below locally, what do you think?

```
diff --git a/overlord/standby/standby_test.go b/overlord/standby/standby_test.go
index 5427814cf..a669a9dbd 100644
--- a/overlord/standby/standby_test.go
+++ b/overlord/standby/standby_test.go
@@ -153,22 +153,44 @@ func (s *standbySuite) TestStartChecks(c *C) {
 
 func (s *standbySuite) TestStopWaits(c *C) {
        defer standby.MockStandbyWait(time.Millisecond)()
-       defer standby.MockStateRequestRestart(func(*state.State, state.RestartType) {})()
+       defer standby.MockStateRequestRestart(func(*state.State, state.RestartType) {
+               c.Fatal("request restart should have not been called")
+       })()
 
        ch := make(chan struct{})
+       opineReady := make(chan struct{})
+       opineGo := make(chan struct{})
        m := standby.New(s.state)
+       synced := false
        m.AddOpinion(opine(func() bool {
+               if !synced {
+                       // synchronize with the main goroutine only at the
+                       // beginning
+                       opineReady <- struct{}{}
+                       <-opineGo
+                       synced = true
+               }
                time.Sleep(200 * time.Millisecond)
                return false
        }))
 
        m.Start()
 
-       // wait enough so the actual Opinionator is called
-       time.Sleep(10 * time.Millisecond)
+       <-opineReady
 
        go func() {
-               m.Stop()
+               stopDone := make(chan struct{})
+
+               // let the opinionator start its delay
+               opineGo <- struct{}{}
+               go func() {
+                       // this will block until standby stops
+                       m.Stop()
+                       close(stopDone)
+               }()
+
+               <-stopDone
+
                close(ch)
        }()
 
@@ -178,4 +200,12 @@ func (s *standbySuite) TestStopWaits(c *C) {
        case <-ch:
                c.Fatal("stop should have blocked and didn't")
        }
+
+       // wait for Stop to complete now
+       select {
+       case <-ch:
+               // nothing to do here
+       case <-time.After(10 * time.Second):
+               c.Fatal("stop did not complete")
+       }
 }
```